Docker astrometry

fergusL opened the pull request:
PR includes changes to add astrometry.net functionality to the huntsman-drp docker container. Also includes a python script that can be pointed at a directory containing fits images to process the astrometry of the files

fergusL committed:
add astrometry.net to huntsman-drp docker container

fergusL committed:
tweaked download_index_files.sh

fergusL committed:
typo

fergusL committed:
tweak

fergusL committed:
tweak dockerfile astrometry install

fergusL committed:
tweak

fergusL committed:
tweak

fergusL committed:
tweak

fergusL committed:
wcslib install tweak

fergusL committed:
tweak

fergusL committed:
tweak

fergusL committed:
tweak

fergusL committed:
tweaks

fergusL committed:
tweak

fergusL committed:
tweak

fergusL committed:
tweak

fergusL committed:
tweak

fergusL committed:
tweak

fergusL committed:
tweak

fergusL committed:
this script needs to be in the docker build context

fergusL committed:
clean up of dockerfile

fergusL committed:
typo  fixes

danjampro commented on scripts/download_index_files.sh:
```diff
@@ -0,0 +1,47 @@
+mkdir -p ~/index_files/
+wget -r -l1 -c -N --no-parent --no-host-directories --cut-dirs=2 -A "index-420[3-9]*.fits" -P ~/index_files/ http://broiler.astrometry.net/~dstn/4200/
+# because I don't know how to regex...``
+wget -r -l1 -c -N --no-parent --no-host-directories --cut-dirs=2 -A "index-4211*.fits" -P ~/index_files/ http://broiler.astrometry.net/~dstn/4200/
```
Missing newline character - you can make atom insert these for you.

danjampro commented on docker/plate_solve_directory.py:
```diff
@@ -0,0 +1,103 @@
+    with Pool(args.nproc) as pool:
+        work = partial(solve_field_task, options)
+        for _ in tqdm.tqdm(pool.imap_unordered(work, fpaths), total=len(fpaths)):
+            pass
```
Missing newline character - you can make atom insert these for you.

danjampro commented on docker/bash_config.sh:
```diff
@@ -6,3 +6,7 @@ eups declare huntsman_drp v1 -r "${LSST_HOME}/huntsman-drp"
+
+alias python=python3
+# need this so panoptes-utils knows where to find solve-field executable
+export SOLVE_FIELD=$(which solve-field)
```
Missing newline character - you can make atom insert these for you.

danjampro commented on docker/astrometry_compile.sh:
```diff
@@ -0,0 +1,30 @@
+#! /bin/bash
+
+# set environment variables:
+# export PKG_CONFIG_PATH=/usr/lib/pkgconfig:/usr/lib64/pkgconfig
```
If we don't need these env variables, I would remove the comments.

danjampro commented on docker/astrometry_compile.sh:
```diff
@@ -0,0 +1,30 @@
+cd astrometry.net
+
+# Edit 'makefile.netpbm' to look like this:
+# NETPBM_INC ?= -I/usr/include/netpbm
```
Remove these comments.

danjampro commented on docker/astrometry_compile.sh:
```diff
@@ -0,0 +1,30 @@
+#! /bin/bash
```
Why do we need a script? Why not just put all this in the dockerfile?

fergusL replied:
Just figured it makes the dockerfile a bit easier to follow. Can just pull everything out the script and put it into the docker file

danjampro commented on docker/astrometry_compile.sh:
```diff
@@ -0,0 +1,30 @@
+cd util
+ORIGINAL_LINE_START='NETPBM_INC.*'
+REPLACE_WITH='NETPBM_INC ?= -I\/usr\/include\/netpbm\/'
+sed -i "s/$ORIGINAL_LINE_START/$REPLACE_WITH/g" makefile.netpbm
```
Is this really the best way of installing astrometry.net? Seems a bit involved.

fergusL replied:
You can install by apt but I have had issues with that in the past.. not sure how reliable it is now. In any case the LSST docker base image is centOS not ubuntu so can't use apt(?) anyway.

danjampro commented on docker/plate_solve_directory.py:
```diff
@@ -0,0 +1,103 @@
+                        to a list of solve-field arguments
+        fpath (str): The filepath of the file to be processed by solve-field
+    """
+    logger = logging.getLogger('main-log')
```
Logging should be handled consistently throughout the project, i.e. via the `HuntsmanBase` class. I'm not sure why you have separated the loggers out into `debug` etc. I think you should be using `logger.debug("message")` instead.

danjampro commented on docker/plate_solve_directory.py:
```diff
@@ -0,0 +1,103 @@
+        logger.warning(f"File does not exist, {fpath}")
+        return
+    if fpath.endswith('.fz'):
+        try:
```
Can you add some comments about what all these `try` `except` statements are trying to achieve?

danjampro commented on docker/plate_solve_directory.py:
```diff
@@ -0,0 +1,103 @@
+    try:
+        get_solve_field(fpath, **options)
+    except Timeout as e:
+        debug_logger.warning(f"get_solve_field() Timed out on {fpath} with error: \n{e}\n")
```
Not sure we need to be adding newline characters for logger messages generally.

fergusL replied:
The error messages that come out of solve_field errors can be several lines long, I find it hard to read the logs without the spaces

wtgee replied:
In general with errors you want to format them `raw`, which should make it more readable:

```suggestion
        debug_logger.warning(f"get_solve_field() Timed out on {fpath} with error: {e!r}")
```

danjampro commented on docker/plate_solve_directory.py:
```diff
@@ -0,0 +1,103 @@
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser()
+    parser.add_argument('--path', type=str, default="./", help="Directory to search for files for processing.")
```
These lines look like they are longer than 100 characters.

danjampro commented on docker/plate_solve_directory.py:
```diff
@@ -0,0 +1,103 @@
+    args = parser.parse_args()
+
+    fpaths = []
+    for dirpath, dirnames, filenames in os.walk(args.path):
```
Add comments for each of these code blocks

danjampro commented on docker/plate_solve_directory.py:
```diff
@@ -0,0 +1,103 @@
+#!/usr/bin/python3
```
I'm not sure what the purpose is of this script - can you add a docstring to the top. Shouldn't it be in the `/scripts` directory?

danjampro commented on docker/plate_solve_directory.py:
```diff
@@ -0,0 +1,103 @@
+            if file.endswith('.fits') or file.endswith('.fits.fz'):
+                fpaths.append(os.path.join(dirpath,file))
+
+    logger = logging.getLogger('main-log')
```
See comments above about logging. Maybe we need to include a `get_logger` function to help us load the logger consistently outside of `HuntsmanBase`.

danjampro commented on scripts/download_index_files.sh:
```diff
@@ -0,0 +1,47 @@
+#! /bin/bash
```
A lot of commented lines here. I'm not sure we need to keep the tutorial documentation inside our own files, but might be nicer to give a link to the documentation. Can you give a short description of why we need these files somewhere?

danjampro commented on docker/dockerfile:
```diff
@@ -25,7 +25,59 @@ RUN yum install -y wget vim && \
+  cd .. && \
+  rm -rf cfitsio/ && \
+  rm -f cfitsio.tar.gz
+
```
In general we want to avoid using multiple `RUN` commands because they can increase the size of the containers. It would be good to move all `yum` installs to a single line, same goes for pip. I'm also not sure why you need to install cfitsio here, as this should be installed already. Same goes for git, gcc, make , fitsio, astropy.

danjampro commented on docker/dockerfile:
```diff
@@ -25,7 +25,59 @@ RUN yum install -y wget vim && \
+
+
 WORKDIR ${LSST_HOME}
+ADD ./plate_solve_directory.py .
```
Why not leave this as a script in `/scripts/astrometry`?

fergusL replied:
you can only specify one docker build context, which atm is `${HUNTSMAN_DRP}/docker`. I think I had it in scripts originally and moved it to be within the build context. If there is no issue with changing the context to just be the entire repo then I can move it back to /scripts/

danjampro reviewed (changes requested):
Hi @fergusL, thanks for your work on this - looks good. I've included a fair few comments, mainly about reorganising / tidying up the code and making things a bit more consistent with the rest of `huntsman-drp`. Happy to discuss any of these with you if they are not clear / don't make sense.

fergusL committed:
danp suggestions1

fergusL committed:
danp suggestions2

fergusL committed:
more formatting changes

fergusL committed:
cleaned up dockerfile, added astrometry build commands to dockerfile

fergusL committed:
moved the process astrometry for whole directory script to scripts/

fergusL committed:
typo

fergusL committed:
dockerfile tweak

fergusL committed:
added symbolic link to lsst install of cfitsio.pc so astrometry.net will compile

fergusL committed:
added `/usr/lib64` to lsst conda env vars 

needed so that the ld installed in the lsst env can find necessary lib files to compile astrometry.net

fergusL committed:
moved the change to LD_LIBRARY_PATH to before activate of env in dfile

fergusL committed:
revert to installing separate gcc outside of conda env

fergusL committed:
tweak

fergusL committed:
added the cfitsio install back in

fergusL committed:
typo

fergusL committed:
pip install some modules outside of conda env

fergusL committed:
typo

fergusL committed:
reverting back to older dockerfile

fergusL committed:
woops

fergusL committed:
reverting to old version of dockerfile

fergusL committed:
asadasd

fergusL committed:
Actually working now...

fergusL committed:
updated readme

fergusL committed:
index data script change

fergusL committed:
consolidated docker RUN blocks

fergusL committed:
Merge branch 'develop' into docker-astrometry

danjampro commented on README.md:
```diff
@@ -4,3 +4,31 @@
 
 # huntsman-drp
 The Huntsman data reduction pipeline (`huntsman-drp`) is responsible for creating calibrated science data from raw images taken by the Huntsman telephoto array. The pipeline uses the LSST code stack configured using the [AstroHuntsman/obs_huntsman](https://github.com/AstroHuntsman/obs_huntsman) package. Please see the [wiki](https://github.com/AstroHuntsman/huntsman-drp/wiki) for instructions on  ...
+
```
I've moved most of the readme stuff into the wiki because its easier to edit. There you can find seconds on testing etc, so a lot of this can be removed. Maybe you can move your description to a new wiki page called "astrometry.net"?

danjampro commented on scripts/astrometry/plate_solve_directory.py:
```diff
@@ -0,0 +1,121 @@
+"""Script to plate solve all fits/fits.fz files in a given directory.
```
Can you add to this to make it clear that it will modify the FITS headers?

danjampro reviewed (commented):
Hi @fergusL, looks good. Just a few minor comments below. Having said that, something in this PR is causing the tests to fail. It looks like something is breaking the `eups` setup of `huntsman-drp` inside the docker container. Once that is fixed I can approve!

fergusL committed:
removed setting of PYTHONPATH env as it conflicts with eups stuff(?)

fergusL committed:
dan changes
